Login

barbarakos opened the pull request:
Added LoginVM, LoginUseCase, LoginDatasource and LoginClient.

barbarakos committed:
Login

Added LoginVM, LoginUseCase, LoginDatasource and LoginClient.

barbarakos committed:
Merge branch 'feature_router' into feature_login

marinzg commented on QuizApp/Login/LoginClient.swift:
```diff
@@ -0,0 +1,71 @@
+        guard let URL = URL(string: "\(baseURL)\(loginPath)") else {
+            throw RequestError.invalidURL
+        }
+        var URLrequest = URLRequest(url: URL)
```
New line after `guard`.

marinzg commented on QuizApp/Login/LoginClient.swift:
```diff
@@ -0,0 +1,71 @@
+
+}
+
+struct LoginRequestModel: Encodable {
+
+    let password: String
+    let username: String
+
+}
+
+struct LoginResponseModel: Decodable {
+
+    let accesToken: String
+
+}
+
+enum RequestError: Error {
+
+    case clientError
+    case serverError
+    case dataError
+    case invalidURL
+    case unknown
+
+}
```
Put models into a separate file.

marinzg commented on QuizApp/Login/LoginDatasource.swift:
```diff
@@ -0,0 +1,12 @@
+    private var loginClient = LoginClient()
+
+    func login(username: String, password: String) async throws -> LoginResponseModel {
+        let response: LoginResponseModel = try await loginClient.login(password: password, username: username)
+        return response
```
Can't you just return from client?
```swift
func login(username: String, password: String) async throws -> LoginResponseModel {
    return try await loginClient.login(password: password, username: username)
}
```

marinzg commented on QuizApp/Login/LoginDatasource.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+class LoginDatasource {
```
For each implementation, you need to have a protocol.
This goes for usecases, repositories, datasources and clients.

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg commented on QuizApp/Login/LoginViewController.swift:
```diff
@@ -16,6 +17,7 @@ class LoginViewController: UIViewController {
         self.init()
 
         self.router = router
+        loginViewModel = LoginViewModel(router: router)
```
You shouldn't create view model in the view controller, it should all be done in the router:
```swift
// Router
let viewModel = MyViewModel(router: Router(), dependency: Dependency())
let viewController = MyViewController(viewModel: viewModel)
// display viewController

// MyViewController
...
init(viewModel: MyViewModel) {
    self.viewModel = viewModel
}


// My ViewModel
...
init(router: Router, dependency: Dependency) {
    self.router = router
    self.dependency = dependency
}
```

marinzg commented on QuizApp/Login/LoginViewController.swift:
```diff
@@ -30,6 +32,7 @@ class LoginViewController: UIViewController {
     }
 
     @objc func handleLogIn() {
+        loginViewModel.login(username: emailTextField.text!, password: passwordTextField.text!)
```
I know you're sure there's going to be something here, but it would be better to unwrap it with guard.

barbarakos committed:
Changes to login.

- models in separate file
- protocols for usecases, datasources and clients
- changes to router

marinzg commented on QuizApp/Login/LoginDatasource.swift:
```diff
@@ -0,0 +1,17 @@
+
+class LoginDatasource {
+
+    private var loginClient = LoginClient()
```
Same thing as for view controller.
All classes should have their dependencies injected through init.

barbarakos committed:
Requested changes
